### PR TITLE
[B-8] Add movie review

### PR DESCRIPTION
**Pull request by DanwerOK**

### CODE REVIEW

File: src/main/java/com/ponodan/movieland/dao/api/ReviewDao.java [hunk 1]
```diff
@@ -1,7 +1,10 @@
 package com.ponodan.movieland.dao.api;
 
 import com.ponodan.movieland.domain.entity.Movie;
+import com.ponodan.movieland.domain.entity.Review;
+import com.ponodan.movieland.domain.entity.UserToken;
 
 public interface ReviewDao {
     void enrichSingleMovieWithReview(Movie movie);
+    int add(Review review, UserToken userToken);
```
Comment (groksana) [inline src/main/java/com/ponodan/movieland/dao/api/ReviewDao.java]: You can add User as attribute in Review and remove UserToken from add

File: src/main/java/com/ponodan/movieland/web/controller/ReviewController.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        stopWatch.start();
+
+        ReviewDto reviewDto = JsonConverter.parseReviewDto(json);
+        Review review = ReviewDtoTransformer.transformDtoToEntity(reviewDto);
+        UserToken userToken = securityService.getTokenByUuid(uuid);
+
+        if (userToken == null){
+            throw new SecurityException("There is no registered UserToken from provided UUID");
+        }
+
```
Comment (groksana) [inline src/main/java/com/ponodan/movieland/web/controller/ReviewController.java]: I think it should be part of security service and not directly in controller

File: src/main/java/com/ponodan/movieland/web/controller/ReviewController.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        boolean isRoleHasWriteAccess = securityService.isRoleHasWriteAccess(userToken.getRole());
+
+        if (isRoleHasWriteAccess){
+            reviewService.add(review, userToken);
+        } else {
+            throw new AccessException("User does not have write access to add review");
+        }
```
Comment (groksana) [inline src/main/java/com/ponodan/movieland/web/controller/ReviewController.java]: I see logic for authorization service here. Don't like such checks in controller directly

File: src/main/java/com/ponodan/movieland/dao/api/ReviewDao.java (hunk 1 above)
Comment (Rooman) [reply src/main/java/com/ponodan/movieland/dao/api/ReviewDao.java]: You should*

File: src/main/java/com/ponodan/movieland/service/impl/ReviewServiceImpl.java [hunk 4]
```diff
@@ -2,17 +2,28 @@
...
+        this.securityService = securityService;
     }
 
     @Override
     public void enrichSingleMovieWithReview(Movie movie){
         reviewDao.enrichSingleMovieWithReview(movie);
     }
+
+    @Override
+    public void add(Review review, UserToken userToken){
```
Comment (juliamanunia) [inline src/main/java/com/ponodan/movieland/service/impl/ReviewServiceImpl.java]: you need to return review with keyHolder id to make it work on UI

File: src/main/java/com/ponodan/movieland/dao/api/ReviewDao.java (hunk 1 above)
Comment (DanwerOK) [reply src/main/java/com/ponodan/movieland/dao/api/ReviewDao.java]: Fixed

File: src/main/java/com/ponodan/movieland/service/impl/ReviewServiceImpl.java (hunk 4 above)
Comment (DanwerOK) [reply src/main/java/com/ponodan/movieland/service/impl/ReviewServiceImpl.java]: Fixed

File: src/main/java/com/ponodan/movieland/web/controller/ReviewController.java (hunk 2 above)
Comment (DanwerOK) [reply src/main/java/com/ponodan/movieland/web/controller/ReviewController.java]: Logic updated according to suggested approach

File: src/main/java/com/ponodan/movieland/web/controller/ReviewController.java (hunk 3 above)
Comment (DanwerOK) [reply src/main/java/com/ponodan/movieland/web/controller/ReviewController.java]: Updated

**Pull request merged**